### ISSUE TITLE
Add setting to menu block for changing breakpoint

### ISSUE THREAD

**Issue by ghost** [backdrop/backdrop-issues#4663]
**Status** [31st Dec]: [PR](https://github.com/backdrop/backdrop/pull/3343) needs (hopefully) final code review.

---

**Description of the need**

When you tick 'Display menu toggle button on small screens' in a menu block, it changes the menu to a menu toggle button on small screens. What seems to determine this 'small screen' size are a set of media query breakpoints in some or all of the following CSS files:

- [`menu-dropdown.css`](https://api.backdropcms.org/api/backdrop/core%21themes%21basis%21css%21component%21menu-dropdown.css/1) (Basis-specific)
- [`menu-toggle.theme.css`](https://api.backdropcms.org/api/backdrop/core%21modules%21system%21css%21menu-toggle.theme.css/1) (theme-independent)
- [`menu-dropdown.theme.css`](https://api.backdropcms.org/api/backdrop/core%21modules%21system%21css%21menu-dropdown.theme.css/1) (theme-independent)

What's frustrating is when you need to change this breakpoint. As far as I know, the only way to do this currently is to copy these CSS files into your theme and then edit the media query breakpoints.

**Proposed solution**

I propose instead a new setting in menu blocks that let you specify a breakpoint to use (preferably a field where you can enter whatever number you need). I'm not sure how this would work under-the-hood, but it'd make for a much better developer/themer experience IMO.

Related:
* https://github.com/backdrop/backdrop-issues/issues/4167 Decide on best way to make improvements to the CSS for core without breaking existing sites

---

Advocate: @BWPanda 

PR: https://github.com/backdrop/backdrop/pull/3343

**Comment by laryn** [comment]
It would expand the scope of what you are suggesting, but I wonder if we should consider adding breakpoint functionality to core as D8 has done. This would allow core to manage breakpoints that could be used by modules, themes, layouts, ...

I started work on porting that module from D7: https://github.com/backdrop-ops/contrib/issues/409

**Comment by klonos** [comment]
Well, we are aiming for as much feature parity with D8 as possible (where it makes sense, and for features that would make people hesitate choosing Backdrop over D8), so I'd be all in for a breakpoints API in core, that both core and contrib could then use.

Breakpoints in core would help with make developing complex frontend themes easier, but could also be used in responsive things in the admin UI, like the admin bar, the primary tabs, the "low priority" columns in tables etc. Breakpoints could even be used as a visibility condition for blocks.

**Comment by BWPanda** [comment]
@laryn Can I ask that breakpoints in core become a seperate issue? Then this one can (or might not) depend on that one.

**Comment by BWPanda** [comment]
I'm really close to getting this working!

**Comment by BWPanda** [comment]
Ok, here's a working PR: https://github.com/backdrop/backdrop/pull/3343

A few caveats though:

- It's not possible to add this setting to menu blocks along with other settings (like enabling the toggle button), because the CSS is applied site-wide, not on a per-block basis.
- I wasn't sure where to put the UI setting (or if we need a UI for it...?) so just left it as a default config value in the `system.core.json` file for now. I.e. Manually set that config value to test/use this feature.
- Basis overrides one of the system menu CSS files with its own, and calls it from its info file. Since I can't get the config value in an info file, I've had to add the breakpoint-specific CSS file in `basis_preprocess_page()`.

The PR basically moves the relevent media query styling from the files mentioned in the OP above to breakpoint-specific CSS files. Then, in each place where the original files were being called, I added a call to the new breakpoint-specific file that includes a media query that uses the min-width value of the new config setting (which can be in 'em', 'px', etc.).

So to test, change the config value to a new min-width (e.g. `config_set('system.core', 'menu_breakpoint', '1200px')`), then test to see if/when the menu changes from a drop-down the a regular menu, and when the menu toggle appears.

Note that when watching the following screencast, '48em' = '768px':
![Peek 2020-10-02 19-51](https://user-images.githubusercontent.com/2385329/94910963-c80ab200-04e8-11eb-8464-9860da485072.gif)

**Comment by stpaultim** [comment]
@BWPanda - Interesting work. I tested this locally and it seems to work as you described. I'm not yet sure how I feel about this. Here are my questions, concerns, and thoughts:

1) I like the idea of looking [broader at breakpoints in core](https://github.com/backdrop/backdrop-issues/issues/4669). This (menu block) feels like it might be a small part of bigger problem and I'm concerned about solving this specific problem in a way that isn't compatible with the larger goal. If we can do this so that's it's compatible with a later solution to the larger goal, that would be great. 

2) I think that the ability to set a breakpoint for the menu would be a nice feature. I'd like to have that. I'd like to have a UI for that. 

3) Where would that UI be? Here are a few of my initial thoughts...

* In a new global settings tab under the APPEARANCE menu?
* configuration/system/mobile/ ?
* configuration/mobile ?

4) While I like the idea, I'm anxious to hear what others think.

**Comment by indigoxela** [comment]
> This (menu block) feels like it might be a small part of bigger problem

That's true, but on the other hand the menu block breakpoints are the ones that most likely need adaption - depending on how many items the main menu is supposed to have and how it is styled.
Whereas, for instance, grid-flexbox.css (layout module) is more likely to fit for different themes/setups - and it feels easier to override.

> Where would that UI be? ... In a new global settings tab under the APPEARANCE menu?

Actually that would be the first place I'd personally search for such a setting. But there might be different opinions.

I agree that there should be a UI for that. Otherwise the menu breakpoints would stay a "hidden feature" only few people know about.

**Comment by BWPanda** [comment]
I've updated the PR to add a UI for configuring the breakpoint:

![image](https://user-images.githubusercontent.com/2385329/95141667-86fef000-07bd-11eb-8bb6-a114c9ff2d89.png)

![image](https://user-images.githubusercontent.com/2385329/95141695-954d0c00-07bd-11eb-9333-cc56e8b230b0.png)

I chose the Menus page to add the form since the Appearance page is for themes, and this setting is specific to menus.

**Comment by indigoxela** [comment]
> I chose the Menus page to add the form since the Appearance page is for themes, and this setting is specific to menus.

Valid point.

The PR works as expected, but I have some suggestions:

Currently the new menu item seems to appear somewhere in the middle in admin menu, at least in the sandbox. Not sure if it should be above or below.
![menu-item-order](https://user-images.githubusercontent.com/842066/95169734-a42ec100-07b3-11eb-9f4f-d8dfe561369f.png)

I think the new form might need some validation. Currently there's none at all.
![no-validation](https://user-images.githubusercontent.com/842066/95169825-c6284380-07b3-11eb-9b7f-b8e8776c4a44.png)

**Comment by BWPanda** [comment]
@indigoxela I think the position of the menu item in the admin menu might be a different issue. From the looks of it, menu items are a bit all over the place:

![image](https://user-images.githubusercontent.com/2385329/95310752-8018d000-08d8-11eb-9250-fb9c3c3d8c51.png)

![image](https://user-images.githubusercontent.com/2385329/95310825-92930980-08d8-11eb-94d5-8c2d0f29726b.png)

Having said that, I've set a weight on the new menu item of '99' that seems to have pushed it to the bottom (though none of the other menu items in that function have weights set).

As for validation, I'm not sure it's going to be possible, or worthwhile. Since you can enter any length in any type of measurement (px, em, %, etc.) then there's no easy way to validate this. I guess we could use some kind of regular expression, but then is it really worth the additional code complexity...?

**Comment by indigoxela** [comment]
>  I think the position of the menu item in the admin menu might be a different issue...

Yeah, seems to be.

> I've set a weight on the new menu item of '99' that seems to have pushed it to the bottom

Works like a charm. Could the menu item weight have any side effects?

> validation ... is it really worth the additional code complexity...?

It's up to you (only a suggestion). Maybe just a simple "plausibility check" to prevent silly typos?

`preg_match('/[0-9.]+[a-z%]+/', $input);`

But possibly you're right. It might not be worth the effort.

**Comment by stpaultim** [comment]
Nice work. I tested it as well. I'm comfortable with the UI. I noticed that you must clear the cache for it to take effect. Should we remind folks of that in help text?

Are there any backwards compatibility issues? Is there any risk that this will break existing sites?

**Comment by BWPanda** [comment]
I added @indigoxela's validation check, @stpaultim's cache flush, and an update hook for existing sites.

**Comment by indigoxela** [comment]
> Are there any backwards compatibility issues? Is there any risk that this will break existing sites?

That's a good question, and an uncomfortable one...

It will have no impact at all, if they never tried to override the core menu breakpoints. It's just that the same value is configurable now.

How about sites/themes that were overriding the relevant css files? The configurable breakpoints ship with new files (so their content won't be overridden, I guess).

**Comment by BWPanda** [comment]
> The configurable breakpoints ship with new files (so their content won't be overridden, I guess).

Hmm, good point...

**Comment by indigoxela** [comment]
>> The configurable breakpoints ship with new files (so their content won't be overridden, I guess).

> Hmm, good point...

What does that mean for this feature request? I hope this doesn't make it a "wontfix", because I think this feature is a big improvement for site builders.

**Comment by BWPanda** [comment]
If there's no way to implement this without breaking existing sites, then this'll become a 2.x feature when we'll be allowing breaking changes. But that's years away, so to get this in in the foreseeable future we need to find a way not to break existing sites...

**Comment by indigoxela** [comment]
> If there's no way to implement this without breaking existing sites, then this'll become a 2.x feature

That would be a shame...

> we need to find a way not to break existing sites

Hm... Brainstorming into the comment form...

We could keep the old css files unchanged and provide alternate new ones. And based on a setting like "Use UI interface to change breakpoints", switch to the new ones (that are otherwise unloaded).

That would make sure that an admin has seen and decided to use the setting. And they hopefully know about how their theme handles it (overrides or leaves untouched).

The downside is a bunch of (almost) duplicate css code and that we sometimes later have to deprecate that use/switch setting.

**Comment by stpaultim** [comment]
> We could keep the old css files unchanged and provide alternate new ones. And based on a setting like "Use UI to change breakpoints", switch to the new ones (that are otherwise unloaded).

The idea of supplemental stylesheets was brought up in relationship to Basis theme, but has also been discussed to address problems in core. I am very uncomfortable with the idea that we can't make css changes to core for several years, if there is the potential to break existing sites. At the same time, I don't want to break existing sites. https://github.com/backdrop/backdrop-issues/issues/4512

**Comment by indigoxela** [comment]
> The idea of supplemental stylesheets was brought up in relationship to Basis theme

Yes, that other issue sort of inspired me to my suggestion. Backwards compatibility isn't always easy to achieve, but still should be possible. Also for core css files.

**Comment by olafgrabienski** [comment]
Coming late to this: Great feature! Trying to improve the help text ...

Current text:

> The minimum width at which menus will be displayed normally. Below this they may be dispayed differently, depending on the menu-specific settings. E.g.: '48em'.

Suggestion:

> The minimum width at which menus will be displayed normally. On smaller screens they may be displayed differently, depending on the menu-specific settings, e.g. as a toggle button. Examples: `48em`, `544px`

**Comment by olafgrabienski** [comment]
> We could keep the old css files unchanged and provide alternate new ones. And based on a setting like "Use UI to change breakpoints", switch to the new ones (that are otherwise unloaded).

Rough UI proposal:

`min-width:`

`The minimum width at which menus will be displayed normally. (...)`

`(·) Default (48em)`
`( ) Custom [        ]`


Default would continue to use the old files.

**Comment by BWPanda** [comment]
I've updated the PR: https://github.com/backdrop/backdrop/pull/3343

New settings form now looks like this:
![image](https://user-images.githubusercontent.com/2385329/95800769-f2a80680-0d43-11eb-81f5-a4116f95c6cc.png)

![image](https://user-images.githubusercontent.com/2385329/95800789-005d8c00-0d44-11eb-903e-e260e89c82ce.png)

**Comment by stpaultim** [comment]
This looks good and generally works for me. However, I'd really like to see the default option list both ems and points. 

> Default (48ems or 768px)

48ems is meaningless to me without additional research, I have a pretty good idea what 768px is off the top of my head.

**Comment by BWPanda** [comment]
> However, I'd really like to see the default option list both ems and points.

Converting `em` to `px` depends upon the text size, so it can change. I'm not sure it's worth adding that to the radio button label...

**Comment by stpaultim** [comment]
Yeah, I see your point. I was thinking that the default value would always be the same, but I suppose it can change depending upon the specific theme. Right? (I don't think it's necessary to try and do a conversion on the fly).

I'm happy with this as it is. Works for me.

**Comment by BWPanda** [comment]
> I suppose it can change depending upon the specific theme. Right?

Yep. In Basis 48em measures 768px, whereas in Bartik 48em measures 672px.

**Comment by indigoxela** [comment]
Almost...
I realized one side effect: as the order of css files seemed to have changed now (?), the hover style of the main menu gets mangled. At least in Basis.

See screenshot (or try it yourself in the sandbox).
![menu-hover-basis](https://user-images.githubusercontent.com/842066/95825008-87086e00-0d30-11eb-9df3-a524428d607c.png)

**Comment by BWPanda** [comment]
Ah ok, think I worked it out. Basis excludes the core dropdown CSS file, but I added two extra core dropdown CSS files for when you choose a custom breakpoint. So I updated the PR to exclude them too. Please try again.

**Comment by indigoxela** [comment]
Ha! Nice catch. Basis is one of those themes that _override_ the system default css files. So it's the perfect proof of concept for us. Updating Backdrop will _not_ break anything in themes.

**Comment by indigoxela** [comment]
Code looks good - RTBC. :+1: 

This is one of those (bigger) PRs that really need another pair of eyes, though.

Anyway, I think it's ready to get a milestone set (by someone else).

**Comment by herbdool** [comment]
Wouldn't it be better to put the setting next to the other hamburger options in the menu block? Just below this screenshot.

![Screenshot_20201016-010855__01](https://user-images.githubusercontent.com/1071483/96215609-7a606180-0f4c-11eb-83db-f27ec1c89dba.jpg)

**Comment by herbdool** [comment]
Apologies, just read https://github.com/backdrop/backdrop-issues/issues/4663#issuecomment-702637373 so menu block won't work. But seems that might be the first place people look. So perhaps a link from the settings block to the global location.

**Comment by BWPanda** [comment]
Good idea! Updated PR:

![image](https://user-images.githubusercontent.com/2385329/96217877-467a5000-0fcf-11eb-9bc8-336df9ccf5d5.png)

**Comment by indigoxela** [comment]
@BWPanda I suspect, there needs to be a permission check for that link. Administer Layouts and Administer menus are different permissions and the latter might not be granted to the user.

**Comment by BWPanda** [comment]
@indigoxela What do you think should happen if someone's editing a menu block who doesn't have the 'Administer menus' permission in order to access the settings page? Should they not see that sentence explaining where the breakpoint can be configured?

As it is now, they'll get an 'Access denied' message, which makes sense. I'm concerned that removing the text based on their permissions might make some people miss knowing there's a new setting that can be configured, who can then ask their admin to give them access to set it for them...

**Comment by indigoxela** [comment]
@bwpanda interesting question...

Presenting a link that leads to 403 doesn't seem right. But I also get your concern re "some people miss knowing there's a new setting".

The text "The breakpoint for small screens can be configured in the global Menu settings." seems OK to me, even if people don't have access to that setting.

Maybe use the same text, but provide the link only if user has permission to access it? How does Backdrop usually do things like that?

**Comment by herbdool** [comment]
Based on @indigoxela suggestion I'm marking this needs work.

**Comment by jenlampton** [comment]
Just a reminder that this changes core css, so would either need to wait intil 2.x in 2025, or depend on https://github.com/backdrop/backdrop-issues/issues/4167. (removing milestone)

**Comment by jenlampton** [comment]
> Basis overrides one of the system menu CSS files with its own, and calls it from its info file. Since I can't get the config value in an info file, I've had to add the breakpoint-specific CSS file in `basis_preprocess_page()`.

If the UI is in core, then the solution shouldn't only exist in basis. If it were, people who are not using Basis will toggle the setting and wonder why nothing is happening. Can we do this in `template_preprocess_page()` instead? Maybe it's there already, checking PR...

Alternatively, we could add the setting only for basis (in the basis theme settings) and add the additional stylesheets only in the basis theme. Doing so would mean people not using Basis wouldn't get the new feature.

**Comment by BWPanda** [comment]
> If the UI is in core, then the solution shouldn't only exist in basis.

@jenlampton The solution doesn't only exist in Basis. The solution is in core, but Basis was overriding it. So to fix Basis, there's an *additional* Basis-specific CSS call in `basis_preprocess_page()`. See the PR for details.

**Comment by jenlampton** [comment]
Ah, I missed that. :)

**Comment by BWPanda** [comment]
This is becoming harder to do, so unassigning myself in case someone else wants to take over.

**Comment by BWPanda** [comment]
> Any PR that changes core css would either need to wait until 2.x in 2025, or depend on #4167.

> In order for this change to be backwards compatible, we can't change any css files that are in core

The PR **doesn't** change core CSS files, it just adds some *new* CSS files to core which are *only* loaded if the new UI setting is activated (which it's not by default). So the only time sites (new or existing) will see any difference is if they specifically activate the new setting.

![image](https://user-images.githubusercontent.com/2385329/97644782-569a3100-1a9f-11eb-870a-fe7b1fee41ef.png)

The only red items in the PR are where I changed CSS from loading all the time to load only when the new setting is set to 'default' (which it is by default, so no breaking changes).

> we'd need to reach a consensus on CSS duplication being "okay"

This is, from my understanding of the recent dicussion, the *only* "problem" with this PR. I don't see how this breaks backwards-compatibility, so unless people have an issue with duplicate CSS (in an effort to avoid breaking backwards-compatibility), then I don't see what the problem is here...

**Comment by indigoxela** [comment]
> Any PR that changes core css would either need to wait until 2.x in 2025

@jenlampton I understand your concern, but belief that this change is fully backwards compatible.

It won't break Basis and wont break other themes, because the feature stays turned off by default and if after turning it on there are unwanted side effects, can easily get turned off again. If it's off, all css is exactly as it currently is.

The only open task at the moment is the link from the Block dialog, which needs a permission check.

So please don't kill this feature. It would be such a great improvement.

**Comment by BWPanda** [comment]
@jenlampton removed the milestone with the reasoning that:

> Any PR that changes core css would either need to wait until 2.x in 2025, or depend on #4167.

But I direct your attention to these commits where core CSS was changed in a similar manner to what we're doing here:
- https://github.com/backdrop/backdrop/commit/ef89b49c38bd806bcde47e79fd747eb1e57d197b
- https://github.com/backdrop/backdrop/commit/d6890f404e74e3c0e88f728977370526dcdfdf45
- https://github.com/backdrop/backdrop/commit/a167f5875811b91de74d2dae1cd4e68157aabef1

On that basis (and the reasons stated in the comments since), I'm adding back the milestone.

**Comment by BWPanda** [comment]
I've updated the PR to make the link to the Menu settings page conditional based on the current user's access.

**Comment by jenlampton** [comment]
> But I direct your attention to these commits

For the record: all the examples provided of previous CSS changes included in core only affect specific admin interfaces, so are not at all comparable to something like this, which would affect every single page on the front-end of the site :)

> So please don't kill this feature. It would be such a great improvement.

There was no intention to kill the feature - we just need to be careful not to break existing Backdrop sites when adding new features. Since we're still debating how it might be possible to do that in #4167 I thought it was prudent not to put the cart before the horse. 

> The PR doesn't change core CSS files, it just adds some new CSS files to core which are only loaded if the new UI setting is activated (which it's not by default). So the only time sites (new or existing) will see any difference is if they specifically activate the new setting.

**This is not correct.** The following steps to reproduce will demonstrate how the current PR will break existing sites:

1) Install Backdrop 1.x
2) Create a Basis sub-theme
3) Override the stylesheet `css/component/menu-dropdown.css` from Basis in your sub theme, by copying the file, and adding it into the subtheme info file as follows:

```
type = theme
name = Basis Subtheme
description = Subtheme of basis for testing purposes
backdrop = 1.x
version = BACKDROP_VERSION

base theme = basis

; Override Basis CSS file
stylesheets[all][] = css/component/menu-dropdown.css
```

4) Change the copied stylesheet as follows:
```
.menu-dropdown {
  position: relative;
}
```
becomes
```
.menu-dropdown {
  position: relative;
  background: pink;
  border: 5px solid red;
}
```
5) enable the Basis subtheme
6) If the subtheme is working, the menu will now look like this:

<img width="1424" alt="Screen Shot 2020-11-07 at 5 14 51 PM" src="https://user-images.githubusercontent.com/397895/98454763-07967080-211d-11eb-9309-cdb3293c2329.png">

7) Apply the patch for the current PR `https://github.com/backdrop/backdrop/pull/3343.patch` (or switch to the appropriate branch)

8) Run `update.php` / clear caches

9) Confirm that all the changes you'd worked so hard on to make your menu look different have been lost due to the latest update:

<img width="1424" alt="Screen Shot 2020-11-07 at 5 19 02 PM" src="https://user-images.githubusercontent.com/397895/98454815-7f649b00-211d-11eb-9714-0633b534602c.png">

**Comment by BWPanda** [comment]
Thanks for the STR @jenlampton. I found and fixed the problem there. Please try again now.

**Comment by BWPanda** [comment]
I'm now advocating for this issue (I was advocating for Tugboat PRs previously). Have updated the OP with current status.

**Comment by indigoxela** [comment]
Here's how I tested:

On a (dev) website with a custom theme - subtheme of Basis - that *overrides* menu-toggle.theme.css as such:

MYTHEME.info:
`stylesheets[all][] = css/menu-toggle.theme.css`

template.php
```
function MYTHEME_css_alter(&$css) {
  // Remove default css files.
  unset($css[backdrop_get_path('module', 'system') . '/css/menu-toggle.theme.css']);
}
```

1. I replaced core with the one from BWPanda/backdrop/tree/issue-4663
2. Ran update.php
3. Checked for menu display changes -- none
4. Then on admin/structure/menu/settings changed the setting to a different value
5. Verified that the attached css files were changed
6. Checked for menu display changes -- none (the theme override still has been respected)
7. Changed the theme code (removed the overrides)
8. Menu breakpoint setting was properly applied (checked with aggregation off and on)

If I get it right, this is exactly what we tried to achieve. Theme override still has precedence over UI settings. And even if something would have looked weird -- I could have changed the setting back to "default" and everything would be in previous state again.

Am I correct so far, or am I missing something? In that case I'd say "works for me".

**Comment by klonos** [comment]
>> So please don't kill this feature. It would be such a great improvement.
>
>There was no intention to kill the feature ...

Yeah. I like this feature too 👍 ...and the direction this is taking seems to be backwards-compatible.

A couple of suggestions:

1. Change the `$form['menu_breakpoint_custom']` textfield to a number field + unit select combo. We wouldn't need any custom validation then (only to reasonable `#min`/`#max` values for the number field I guess), and also wouldn't need the help text either:

   ![image](https://user-images.githubusercontent.com/2423362/99670332-0cdab000-2a79-11eb-9662-96ca10144f43.png)

2. This help text is confusing (also has a typo: `dispayed` -> `displayed`), but I have no suggestions at the moment:

   > The minimum width at which menus will be displayed normally. Below this they may be dispayed differently, depending on the menu-specific settings.

**Comment by indigoxela** [comment]
> Change the $form['menu_breakpoint_custom'] textfield to a number field + unit select combo. We wouldn't need any custom validation then (only to set reasonable #min/#max values for the number field I guess)

Hm, are we using such a combo anywhere else in Backdrop? It seems a bit weird.

Furthermore, this would need a process callback, as the config value needs to get combined from two different form items then. IMO this causes more complexity and work than it's worth it.

Another note: your mockup provides a full list of all available CSS units. But media queries for screen can only use two of them: **px** and **em**, potentially rem. At least as far as I know. Here's an article about it: https://zellwk.com/blog/media-query-units/ (didn't find official docs, though)

**Comment by klonos** [comment]
> Hm, are we using such a combo anywhere else in Backdrop? It seems a bit weird.

Yes, we are. See `admin/structure/types/manage/post/fields/field_image`:

![image](https://user-images.githubusercontent.com/2423362/99674382-a3f63680-2a7e-11eb-9a98-bdd1a3397499.png)

> ...media queries for screen can only use two of them

Fair enough, we should only allow those that make sense then.

**Comment by indigoxela** [comment]
Regarding "same as in field cardinality" - I doubt that this is really the same, but am too lazy to proof it. :smile:

**Comment by BWPanda** [comment]
> Change the $form['menu_breakpoint_custom'] textfield to a number field + unit select combo.

I agree with @indigoxela that this might be overly complex. Let's see what others think...

> This help text is confusing

I've updated it to: "The 'min-width' for the menu breakpoint. Screens smaller than this may display a menu toggle button, depending on the menu-specific settings." Is that better?

> I've left a comment with code suggestion on how to conditionally add the link to the global menu settings in a way that it's better for translators.

I left my own comment explaining how I don't think that'll work in this situation :slightly_smiling_face:

**Comment by klonos** [comment]
> ...this might be overly complex. Let's see what others think

👍 

> Is that better?

Yup, definitely better 👍 ...lets wait for others to suggest alternatives/improvements.

> I left my own comment explaining how I don't think that'll work in this situation 🙂

I see no point in showing users something that they cannot action, but yup, need more feedback on [this](https://github.com/backdrop/backdrop/pull/3343#discussion_r527381641):

> @klonos The problem with your suggestion is that users without the `administer menu` permission will never see this text (which might be helpful for them). The idea with my code was that they'd still see the text, it just wouldn't be linked to an 'Access denied' page if they didn't have permission.

**Comment by olafgrabienski** [comment]
This is really close, let's try to get it ready. Trying to give some useful feedback to open questions:

**1) Breakpoint help text**

The last suggestion is good to go with, in my opinion:

> The 'min-width' for the menu breakpoint. Screens smaller than this may display a menu toggle button, depending on the menu-specific settings.

The *description* text below the min-width field could be improved (less complicated, give also an `em` example, cf. default setting):

Current:
  > Enter a valid CSS width value; a number followed by a unit (e.g.: '768px').

Proposal:
  > A valid CSS width value, e.g. `544px` or `34em`.

**2) Change textfield to number + unit combo**

Too complex, and it doesn't really help in my opinion.

**3) Menu block help text**

> The idea with my code was that they'd still see the text, it just wouldn't be linked to an 'Access denied' page if they didn't have permission.

Sounds reasonable to me. That way, people without permission are at least informed and can ask their site architect or developer.

**Comment by klonos** [comment]
> Too complex, and it doesn't really help in my opinion.

Having a number field + dropdown with CSS units helps in the following ways:
- DX:
  - prevents wrong user input, by reusing in-built field validation, so it inherits any future improvements
  - no need for custom `_validate()` function (less custom code -> better DX)
  - no need for complex regex to match all possible user input
- UX:
  - provides "guided input"
  - no need for help text to explain what the expected value is
  - certain HTML5 fields have better support on mobile devices (number fields for example automatically get a numeric keyboard)

> That way, people without permission are at least informed and can ask their site architect or developer.

Noting that this help text is shown in the admin UI that allows tweaking blocks/layouts, so the people that will see it must already be site builders.

**Comment by indigoxela** [comment]
> Having a number field + dropdown with CSS units helps...

@klonos your arguments are valid. But the concerns here are that adding a lot of form complexity might not be worth the small UI/UX benefit.

Setting the menu breakpoint usually isn't an end-user task, but one for a site builder / site admin. IMO they do not need "guided input" to that extent.

**Comment by olafgrabienski** [comment]
@klonos As user of the form I'd prefer one text field: Just typing `62em` or `992px` seems less complicated to me than using two form elements. This is my personal opinion. I think however, we don't *need* a very sophisticated form here.

I'm fine with both approaches. Let's go with what is better implementation wise. (I'm not able to evaluate DX and the technical aspects of the implementation.)

**Comment by findlabnet** [comment]
What about single drop-down with set of predefined standard viewports? No matter which units is used, nothing to validate.

**Comment by klonos** [comment]
That would be too limiting @findlabnet, as we'd need to hardcode value/unit pairs ...either that, or the dropdown would be HUGE.

**Comment by findlabnet** [comment]
OK, I just thought that extra-small, small, medium, large and extra-large should be enough.

**Comment by klonos** [comment]
Yup, makes sense; but extra-small, small, medium, large, and extra-large may have different values from site to site and developer to developer. Right?

**Comment by klonos** [comment]
Anyway, I've provided plenty of feedback here and in the PR, so I'll let others chime in, and core committers decide.

I'll close by saying that it is known that people don't read help text, so it should be avoided when possible, and the UI should be intuitive enough for users to know what to do. To that end, here's my take on the ideal UI for this form:

![image](https://user-images.githubusercontent.com/2423362/101656239-286b2200-3a4b-11eb-891f-94da9d1d532d.png)

PS: this also saves a single property in the config file, so instead of this pair:

```
"menu_breakpoint": "custom",
"menu_breakpoint_custom": "480px"
```

...it's just:

```
"menu_breakpoint": "480px"
```

And instead of this:

```
"menu_breakpoint": "default",
"menu_breakpoint_custom": ""
```

...it's just:

```
"menu_breakpoint": "48em"
```

**Comment by olafgrabienski** [comment]
@klonos How would the default state look like in your mockup?

**Comment by klonos** [comment]
![image](https://user-images.githubusercontent.com/2423362/101657695-da571e00-3a4c-11eb-822b-94911560aae9.png)

The "reset" link is only shown if the value is NOT 48em 😉

**Comment by findlabnet** [comment]
> may have different values from site to site and developer to developer. Right?

I thought it was determined by the screen size of the devices produced / used.
And, for example, in some cases CSS variables can be used to fine tune.

**Comment by olafgrabienski** [comment]
>> How would the default state look like in your mockup?

> The "reset" link is only shown if the value is NOT 48em

Thanks, interesting! I'm also fine with this approach, if the implementation is easy. IIRC, the type of the form is the last open question. Let's wait what @BWPanda says.

**Comment by BWPanda** [comment]
@findlabnet I build responsive sites by designing first for desktop, then I start to reduce the width of my browser until something breaks or doesn't look right. I then add a CSS media query to fix that (generally for that width and smaller - e.g. converting a 3 column grid to 2 columns, etc.). I end up with a site that looks and works nicely for all screen sizes. As such, there's no such thing as 'standard' viewport sizes for me.

Now I'm not saying we should build this new functionality with only me in mind, but this issue started out of a frustration of the hard-coded value Backdrop currently uses. If we 'fix' this by adding a few more hard-coded values, then I don't see that as a good, long-term fix. We'll just be back here again in future with people asking for even more options.

@klonos I guess my point of view is that without this change users would have to overwrite the media query (as @olafgrabienski had to do). In which case they're used to writing CSS widths. This setting just adds that same thing to the UI, instead of needing to overwite CSS files manually. So I figured users would be fine writing out `540px` in a textfield.

Having said that, I'm not totally opposed to the idea if it helps get this feature in. Having said *that*, it seems that there are more people opposed to the combo fields than for them...

**Comment by BWPanda** [comment]
I've updated the description text in the PR based on recent feedback.

**Comment by BWPanda** [comment]
Regarding the combo field option; let's push that to a follow-up issue so as not to hold up this feature for the next release.

**Comment by quicksketch** [comment]
I think we should hold-off on the combo field as well. Though I think we should be more strict about the allowed values to prevent someone from typing in a value that isn't valid. I posted more feedback in a code review: https://github.com/backdrop/backdrop/pull/3343#pullrequestreview-550998385

This looks great @BWPanda! :+1:

**Comment by BWPanda** [comment]
Thanks for the review @quicksketch! I've made a bunch of changes based on your suggestions.

**Comment by quicksketch** [comment]
Implementation looks great to me.

One thing I'd appreciate is switching the default theme to Bartik and visually ensuring that the breakpoint functionality works in a non-Basis theme.

**Comment by BWPanda** [comment]
@quicksketch I tested Bartik and it worked, however it had styling applied to the menu when the screen was >768px (hard-coded). So I had to add a fix for that, so it applies to whatever the custom breakpoint is. Working better now.

There was also a typo that caused the site to break, so I fixed that too. Please re-test/review.

**Comment by stpaultim** [comment]
I haven't tested this in a long while. I took a look tonight and it looks good.

**Comment by indigoxela** [comment]
Another testing round and bad news...

@BWPanda I tested your branch with:

- Basis (works)
- Pelerine (works)
- Tatsu

With Tatsu there's a visual regression regarding main menu hover and also regarding some margins in header. Both, when the setting is default _and_ custom. Flushing caches didn't help.

![header-regression](https://user-images.githubusercontent.com/842066/102187776-b9f7eb00-3eb4-11eb-9204-c9ff92c043ee.png)

I belief, that's because of the changed css file order (Tatsu's skin.css?), but I'm not 100% sure. @stpaultim mind to further investigate?

**Comment by indigoxela** [comment]
Ooops and some CSS related tests are failing. Back to "needs work".

**Comment by stpaultim** [comment]
> I belief, that's because of the changed css file order (Tatsu's skin.css?), but I'm not 100% sure. @stpaultim mind to further investigate? Or is this really supposed to look that ugly?

Well, my intention was to avoid ugly, but I'll do some checking/testing.

**Comment by BWPanda** [comment]
@indigoxela The Tatsu issue is unrelated to this PR. I had the same issue on a demo site: https://1-17-4-qxddvgjiijb6zec5xieztsr1apxxut7q.tugboat.qa/

As for the CSS tests failing, I looked into them but couldn't work out what's going on (I hate how I just can't understand/write/debug tests). So need help with that from someone please.

**Comment by indigoxela** [comment]
> The Tatsu issue is unrelated to this PR.

Good to know.

> As for the CSS tests failing, I looked into them but couldn't work out what's going on

Pretty annoying is that now the "Details" links of the tests lead to 404. So I ran some random tests in the sandbox.

This is one that has failures:

Menu router
Tests menu router and hook_menu() functionality.
219 passes, 4 fails, 0 exceptions, and 46 debug messages.

For example: "FAIL: The maintenance theme's CSS appears on the page."

I'd volunteer to do some local tests, which would be much easier for me if you rebased your branch (it's currently 66 commits behind backdrop:1.x.).

**Comment by BWPanda** [comment]
Thanks @indigoxela, PR updated. Have at it 😄

**Comment by indigoxela** [comment]
It should be an easy fix.

The failing tests search for Bartik's theme.css, but fail to find it, for instance:

`$this->assertRaw('bartik/css/style.css', "The default theme's CSS appears on the page.");`

See my comment for a suggestion to fix, or to dig deeper.

The failing test re config sync... hook_config_info() :rofl: (no comment):

> The configuration "menu.settings" is not owned by any module.

Additionally I'd suggest to switch the logic of your checks for the setting: only act if `setting == 'custom'` (instead of acting if setting != 'default').

**Comment by BWPanda** [comment]
> The failing tests search for Bartik's theme.css, but fail to find it

Ugh, this is why I hate tests - they're so confusing. If a test fails saying "The default theme's CSS appears on the page.", then I take that to mean the CSS was there but shouldn't be. Not the other way around.

Anyway, thanks for looking at this for me @indigoxela. I've updated the PR as per your suggestions and now tests are passing again.

**Comment by klonos** [comment]
> ...then I take that to mean the CSS was there but shouldn't be. Not the other way around.

Haha, took me a while to get my head around that too 😅

**Comment by indigoxela** [comment]
> I've updated the PR as per your suggestions and now tests are passing again.

@BWPanda Great job, many thanks! Code looks good to me now.

Maybe one day you'll be able to make friends with Simpletest - or how I call it "Could-be-worse-test" :smile: Dealing with those tests is often time-consuming, but they're useful.

**Comment by indigoxela** [comment]
Another round of testing, I installed following themes in the sandbox:

- Bootstrap Lite
- Colihaut
- Corporate KISS
- Lateral
- Pelerine
- Shasetsu
- Snazzy

The first thing I checked it that they're all untouched when the setting is "default". :+1: No problems, works as expected.

Then I set a custom breakpoint and checked them all again. Hm, not all are compatible (including mine), but others are. Which makes me think that we could provide a short help text on top of the settings form, saying that not all (contrib/custom) themes may be compatible yet with that new setting. Or is this overdone?

**Comment by BWPanda** [comment]
I think that's technically the same with any UI setting. Any contrib theme can override a setting or ignore it completely...

**Comment by klonos** [comment]
I agree with @BWPanda ...we should not overthink this. If people start using custom breakpoints with their contrib themes, they'll realize that things aren't working as expected, and I'd expect them to start raising issues in the respective issue queues of these contrib themes.

**Comment by indigoxela** [comment]
> ...we should not overthink this

Let's call it a day then, works for me.

**Comment by quicksketch** [comment]
I did a code review and the only substantial thing I found was an issue with removing all of Bartik's `style.css` file. Perhaps this makes sense in a way I'm not understanding, but it seems like it would cause a problem: https://github.com/backdrop/backdrop/pull/3343#pullrequestreview-560563482

**Comment by BWPanda** [comment]
@quicksketch I replied to your concern: https://github.com/backdrop/backdrop/pull/3343#discussion_r550716428

**Comment by quicksketch** [comment]
In reviewing the first time, I hadn't noticed that there is a complete copy of Bartik's `style.css` file that is being duplicated. That's 1700 lines of duplicated CSS just so that we can change a few lines if menu breakpoints are enabled. While the swapping `menu-dropdown.theme.css` file in Basis makes sense because it's dealing exclusively with menus, I don't think that same approach makes sense for Bartik, where the entire `style.css` file has to be swapped out.

As a compromise to get this completed but not have this make a mess of Bartik, I would be in favor of removing the duplicated `style.css`, having this feature simply not work properly in Basis for now, and create a follow-up issue to sort it out separately.

**Comment by olafgrabienski** [comment]
> having this feature simply not work properly in Basis for now

Not working in *Basis* doesn't sound good! Maybe a typo (meaning "not work properly in **Bartik**")?

**Comment by BWPanda** [comment]
@quicksketch Done. Bartik changes removed. PR updated.

@olafgrabienski Yes, I assume that's what @quicksketch meant.

**Comment by BWPanda** [comment]
Here's the follow-up issue for Bartik: https://github.com/backdrop/backdrop-issues/issues/4852

**Comment by quicksketch** [comment]
I think this is good to go. I'd love any last rounds of feedback as to whether we should hold this up or push it through for 1.18.0. Code freeze is tomorrow and we can still merge this in before the deadline.

**Comment by stpaultim** [comment]
I'm comfortable pushing this through for 1.18.0.

**Comment by BWPanda** [comment]
Please merge! I don't want my ability to advocate for other issues to be held up for another 4 months! 😄

**Comment by olafgrabienski** [comment]
Same here, let's get it in 1.18.0 (and mention the state in the release notes).

**Comment by quicksketch** [comment]
Super, thank so much everyone! This was a big effort. I've merged https://github.com/backdrop/backdrop/pull/3343 into 1.x for 1.18.0. A preview release will be coming out later this evening with this included.
